crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,742 @@
+            if line.indent_level < nesting_indent {
+                if line.is_comment_only {
+                    self.fn_status = Status::CommentAfter(nesting_indent);
+                } else {
+                    self.fn_status = Status::Outside;
+                }
+            }
+        }
+
+        // A comment can be de-indented while still being in a class/function, in that case
+        // we need to revert the variables.
+        if !line.is_comment_only {
+            if let Status::CommentAfter(indent) = self.fn_status {
+                if line.indent_level >= indent {
+                    self.fn_status = Status::Inside(indent);
+                }
+                self.fn_status = Status::Outside;
+            }
+
+            if let Status::CommentAfter(indent) = self.class_status {
+                if line.indent_level >= indent {
+                    self.class_status = Status::Inside(indent);
+                }
+                self.class_status = Status::Outside;
+            }
+        }
+
+        // Don't expect blank lines before the first non comment line.
+        if self.is_not_first_logical_line {
+            if line.preceding_blank_lines == 0
+                // Only applies to methods.
+            && line.first_token == Tok::Def
+                && matches!(self.class_status, Status::Inside(_))
+                // The class/parent method's docstring can directly precede the def.
+                && !matches!(self.follows, Follows::Docstring)
+                // Do not trigger when the def follows an if/while/etc...
+                && prev_indent_level.is_some_and(|prev_indent_level| prev_indent_level >= line.indent_level)
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !matches!(self.follows, Follows::Decorator)
+            {
+                // E301
+                let mut diagnostic = Diagnostic::new(
+                    BlankLineBetweenMethods(line.preceding_blank_lines),
+                    line.first_token_range,
+                );
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist.line_ending().as_str().to_string(),
+                    locator.line_start(self.last_non_comment_line_end),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.preceding_blank_lines < BLANK_LINES_TOP_LEVEL
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !matches!(self.follows, Follows::Decorator)
+                // Allow groups of one-liners.
+                && !(matches!(self.follows, Follows::Def) && !matches!(line.last_token, Tok::Colon))
+                // Only trigger on non-indented classes and functions (for example functions within an if are ignored)
+                && line.indent_level == 0
```

Comment:
Nit
```suggestion
                // Only apply to functions or classes.
                && is_top_level_token_or_decorator(&line.first_token)
```

Reply:
I also assumed this was about visual indentation, since the suggestion doesn't make sense (I think ?)

1831879a2